File: bus/sdk/js/caller.js
// SPDX-License-Identifier: MIT

"use strict";

const ACTIONS = Object.freeze(["list", "send", "spawn", "describe", "run", "start", "wait", "status", "interrupt", "close", "forget", "ack"]);
const ACTION_METHODS = Object.freeze({ list: "session.list", send: "message.send", spawn: "lane.spawn", describe: "lane.describe", run: "turn.run", start: "turn.start", status: "turn.status", wait: "turn.wait", interrupt: "turn.interrupt", close: "session.close", forget: "session.close" });

// Run references and results belong to the resident worker, not this caller.
class Caller {
  constructor(connection) { this.connection = connection; }
  /**
   * @param {import("./protocol").SessionListRequest} [request]
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").SessionListResult>}
   */
  list(request = {}, cancel) { return this.connection.call("session.list", request, cancel); }
  /**
   * @param {import("./protocol").MessageSendRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").MessageSendResult>}
   */
  send(request, cancel) { return this.connection.call("message.send", request, cancel); }
  /**
   * @param {import("./protocol").LaneDescribeRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").LaneDescribeResult>}
   */
  describe(request, cancel) { return this.connection.call("lane.describe", request, cancel); }
  /**
   * @param {import("./protocol").LaneSpawnRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").LaneSpawnResult>}
   */
  spawn(request, cancel) { return this.connection.call("lane.spawn", request, cancel); }
  /**
   * @param {string} sessionID
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").LaneSpawnResult>}
   */
  resume(sessionID, cancel) { return this.spawn({ resume_session_id: sessionID }, cancel); }
  /**
   * @param {import("./protocol").TurnRunRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").RunStatus>}
   */
  run(request, cancel) { return this.connection.call("turn.run", request, cancel); }
  /**
   * @param {import("./protocol").TurnRunRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").RunRef>}
   */
  start(request, cancel) { return this.connection.call("turn.start", request, cancel); }
  /**
   * @param {import("./protocol").ReadRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").RunStatus>}
   */
  status(request, cancel) { return this.connection.call("turn.status", request, cancel); }
  /**
   * @param {import("./protocol").WaitRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<import("./protocol").RunStatus>}
   */
  wait(request, cancel) { return this.connection.call("turn.wait", request, cancel); }
  /**
   * @param {import("./protocol").SessionTarget} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<Record<string, never>>}
   */
  interrupt(request, cancel) { return this.connection.call("turn.interrupt", request, cancel); }
  /**
   * @param {import("./protocol").SessionCloseRequest} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<Record<string, never>>}
   */
  close(request, cancel) { return this.connection.call("session.close", request, cancel); }
  // Once submitted, acknowledgement settles consumption despite later abort.
  /**
   * @param {import("./protocol").RunRef} request
   * @param {AbortSignal} [cancel]
   * @returns {Promise<Record<string, never>>}
   */
  ack(request, cancel) {
    if (cancel?.aborted) return Promise.reject(cancel.reason || new Error("aborted"));
    return this.connection.call("turn.ack", request);
  }
  action(name, args = {}, cancel) {
    if (name === "ack") return this.ack(args, cancel);
    const method = ACTION_METHODS[name];
    if (method) return this.connection.call(method, name === "forget" ? { ...args, forget: true } : args, cancel);
    return Promise.reject(new Error("unknown action"));
  }
  disconnected() {}
}

module.exports = { ACTIONS, Caller };

File: bus/sdk/go/caller.go
// SPDX-License-Identifier: MIT

package sessionkit

import (
	"context"
	"encoding/json"

	"github.com/antst/sessionbus/bus/sdk/go/protocol"
)

type callFunc func(context.Context, string, any, any) error
type Caller struct{ call callFunc }
type WaitRequest = protocol.WaitRequest

func newCaller(call callFunc) *Caller {
	return &Caller{call: call}
}

func NewCaller(call func(context.Context, string, any) (json.RawMessage, error)) *Caller {
	return newCaller(func(ctx context.Context, method string, params, result any) error {
		raw, err := call(ctx, method, params)
		if err != nil {
			return err
		}
		return protocol.UnmarshalResult(method, raw, result)
	})
}

func callAs[T any](ctx context.Context, call callFunc, method string, params any) (T, error) {
	var result T
	_, err := protocol.EncodeParams(method, params)
	if err == nil {
		err = call(ctx, method, params, &result)
	}
	return result, err
}

func (c *Caller) List(ctx context.Context, request SessionListRequest) (SessionListResult, error) {
	return callAs[SessionListResult](ctx, c.call, "session.list", request)
}
func (c *Caller) Send(ctx context.Context, request MessageSendRequest) (MessageSendResult, error) {
	return callAs[MessageSendResult](ctx, c.call, "message.send", request)
}
func (c *Caller) Describe(ctx context.Context, request LaneDescribeRequest) (LaneDescribeResult, error) {
	return callAs[LaneDescribeResult](ctx, c.call, "lane.describe", request)
}
func (c *Caller) Spawn(ctx context.Context, request LaneSpawnRequest) (LaneSpawnResult, error) {
	return callAs[LaneSpawnResult](ctx, c.call, "lane.spawn", request)
}
func (c *Caller) Resume(ctx context.Context, sessionID string) (LaneSpawnResult, error) {
	return c.Spawn(ctx, LaneSpawnRequest{ResumeSessionID: sessionID})
}
func (c *Caller) Run(ctx context.Context, request TurnRunRequest) (RunStatus, error) {
	return callAs[RunStatus](ctx, c.call, "turn.run", request)
}
func (c *Caller) Interrupt(ctx context.Context, request SessionTarget) error {
	_, err := callAs[struct{}](ctx, c.call, "turn.interrupt", request)
	return err
}
func (c *Caller) Close(ctx context.Context, request SessionCloseRequest) error {
	_, err := callAs[struct{}](ctx, c.call, "session.close", request)
	return err
}

func (c *Caller) Start(ctx context.Context, request TurnRunRequest) (RunRef, error) {
	return callAs[RunRef](ctx, c.call, "turn.start", request)
}
func (c *Caller) Status(ctx context.Context, request ReadRequest) (RunStatus, error) {
	return callAs[RunStatus](ctx, c.call, "turn.status", request)
}
func (c *Caller) Wait(request WaitRequest) (RunStatus, error) {
	return c.WaitContext(context.Background(), request)
}

// WaitContext abandons only this non-consuming read, never the worker's result.
func (c *Caller) WaitContext(ctx context.Context, request WaitRequest) (RunStatus, error) {
	return callAs[RunStatus](ctx, c.call, "turn.wait", request)
}

// Ack commits consumption. Once submitted, its response wins later cancellation.
func (c *Caller) Ack(ctx context.Context, request RunRef) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	_, err := callAs[struct{}](context.WithoutCancel(ctx), c.call, "turn.ack", request)
	return err
}
func (c *Caller) disconnected() {} // Results belong to the resident worker.
